stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -820,7 +820,7 @@ function eigen(D::Diagonal; permute::Bool=true, scale::Bool=true, sortby::Union{
             evecs[p[i],i] = one(Td)
         end
     else
-        evecs = Matrix{Td}(I, size(D))
+        evecs = Diagonal{eltype(λ)}(ones(length(λ)))
```

Comment:
By moving the type statement, you make sure that you only allocate one vector and don't need to convert the Float64 `one` to something else, to which it may not be convertible to (like in the failing tests):

```suggestion
        evecs = Diagonal(ones(eltype(λ), length(λ)))
```